File: ios/Classes/GrovsPlugin.swift
import Flutter
import UIKit
import Grovs

extension Array where Element == Any {
    func toArray<T>(of type: T.Type) -> [T] {
        return self.compactMap { $0 as? T }
    }
}

extension Array where Element == Any {
    static func convertClosure<T>(
        _ closure: @escaping ([Any]) -> Void,
        to type: T.Type
    ) -> ([T]) -> Void {
        return { typedArray in
            let anyArray = typedArray.map { $0 as Any }
            closure(anyArray)
        }
    }
    
    /// Wraps a `([Any]) -> Void` closure as a `([T]?) -> Void` closure.
    static func convertClosure<T>(
        _ closure: @escaping ([Any]) -> Void,
        toOptionalArrayOf type: T.Type
    ) -> ([T]?) -> Void {
        return { typedArray in
            let anyArray: [Any] = typedArray?.map { $0 } ?? []
            closure(anyArray)
        }
    }
}

/// Forwards native SDK errors to Dart, buffering the ones raised before Dart subscribes.
private final class ErrorStreamHandler: NSObject, FlutterStreamHandler {
    private var sink: FlutterEventSink?
    private var pending: [[String: Any]] = []
    private let maxPending = 20

    func send(_ event: [String: Any]) {
        if let sink = sink {
            sink(event)
            return
        }
        pending.append(event)
        if pending.count > maxPending {
            pending.removeFirst()
        }
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        pending.forEach { events($0) }
        pending.removeAll()
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}

public class GrovsPlugin: NSObject, FlutterPlugin {
    private enum PendingLaunchLink {
        case userActivity(NSUserActivity)
        case url(URL)
    }

    private var eventSink: FlutterEventSink?
    private var methodChannel: FlutterMethodChannel?
    private let errorStream = ErrorStreamHandler()

    // Consent state and the launch link received while disabled are shared by
    // every engine because the SDK is configured once per process.
    private static var sdkEnabled = true
    private static var pendingLaunchLink: PendingLaunchLink?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "grovs", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "grovs/deeplinks", binaryMessenger: registrar.messenger())
        
        let instance = GrovsPlugin()
        instance.methodChannel = channel
        
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
        eventChannel.setStreamHandler(instance)
        let errorChannel = FlutterEventChannel(name: "grovs/errors", binaryMessenger: registrar.messenger())
        errorChannel.setStreamHandler(instance.errorStream)
    }

    // Hook into didFinishLaunchingWithOptions
    public func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [AnyHashable : Any] = [:]) -> Bool {
        // Read API key and test environment flag from Info.plist
        if let infoDictionary = Bundle.main.infoDictionary, let apiKey = infoDictionary["GrovsApiKey"] as? String {
            let useTestEnvironment = infoDictionary["GrovsUseTestEnvironment"] as? Bool ?? false
            let baseURL = infoDictionary["GrovsBaseURL"] as? String
            let clipboardDomains = infoDictionary["GrovsClipboardDomains"] as? [String]
            let enabled = infoDictionary["GrovsEnabled"] as? Bool ?? true
            GrovsPlugin.sdkEnabled = enabled
            Grovs.configure(
                APIKey: apiKey,
                useTestEnvironment: useTestEnvironment,
                baseURL: baseURL,
                autoTrackScreenViews: false,
                clipboardDomains: clipboardDomains,
                enabled: enabled,
                delegate: self
            )
        }
        
        return true
    }
    
    // Handle universal link continuation
    public func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]) -> Void) -> Bool {
        if !GrovsPlugin.sdkEnabled {
            GrovsPlugin.pendingLaunchLink = .userActivity(userActivity)
        }
        return Grovs.handleAppDelegate(continue: userActivity, restorationHandler: Array.convertClosure(restorationHandler, toOptionalArrayOf: UIUserActivityRestoring.self))
    }

    // Handle URI opening
    public func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
        if !GrovsPlugin.sdkEnabled {
            GrovsPlugin.pendingLaunchLink = .url(url)
        }
        return Grovs.handleAppDelegate(open: url, options: options)
    }

    private func setSDKEnabled(_ enabled: Bool) {
        guard enabled != GrovsPlugin.sdkEnabled else { return }
        Grovs.setSDK(enabled: enabled)
        GrovsPlugin.sdkEnabled = enabled

        guard enabled, let pending = GrovsPlugin.pendingLaunchLink else { return }
        GrovsPlugin.pendingLaunchLink = nil
        // The native SDK drops links received while disabled; hand it the launch
        // link again now that it may process it. Must run on the main thread.
        switch pending {
        case .userActivity(let activity):
            _ = Grovs.handleAppDelegate(continue: activity, restorationHandler: { _ in })
        case .url(let url):
            _ = Grovs.handleAppDelegate(open: url, options: [:])
        }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            
        case "configure":
            break
            
        case "generateLink":
            guard let args = call.arguments as? [String: Any],
                  let title = args["title"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "title is required", details: nil))
                return
            }
            
            let subtitle = args["subtitle"] as? String
            let imageURL = args["imageURL"] as? String
            let data = args["data"] as? [String: Any]
            let tags = args["tags"] as? [String]
            let customRedirectsMap = args["customRedirects"] as? [String: [String: Any]]
            let showPreviewIos = args["showPreviewIos"] as? Bool
            let showPreviewAndroid = args["showPreviewAndroid"] as? Bool
            let copyToClipboardIos = args["copyToClipboardIos"] as? Bool
            let copyToClipboardAndroid = args["copyToClipboardAndroid"] as? Bool
            let trackingMap = args["tracking"] as? [String: String]
            
            // Parse custom redirects
            var customRedirects: CustomRedirects?
            if let redirectsMap = customRedirectsMap {
                var ios: CustomLinkRedirect?
                var android: CustomLinkRedirect?
                var desktop: CustomLinkRedirect?
                
                if let iosMap = redirectsMap["ios"] {
                    ios = CustomLinkRedirect(
                        link: iosMap["url"] as? String ?? "",
                        openAppIfInstalled: iosMap["openAppIfInstalled"] as? Bool ?? true
                    )
                }
                if let androidMap = redirectsMap["android"] {
                    android = CustomLinkRedirect(
                        link: androidMap["url"] as? String ?? "",
                        openAppIfInstalled: androidMap["openAppIfInstalled"] as? Bool ?? true
                    )
                }
                if let desktopMap = redirectsMap["desktop"] {
                    desktop = CustomLinkRedirect(
                        link: desktopMap["url"] as? String ?? "",
                        openAppIfInstalled: desktopMap["openAppIfInstalled"] as? Bool ?? true
                    )
                }
                
                customRedirects = CustomRedirects(ios: ios, android: android, desktop: desktop)
            }
            
            // Extract tracking parameters
            let trackingCampaign = trackingMap?["utm_campaign"]
            let trackingSource = trackingMap?["utm_source"]
            let trackingMedium = trackingMap?["utm_medium"]
            
            Grovs.generateLink(
                title: title,
                subtitle: subtitle,
                imageURL: imageURL,
                data: data,
                tags: tags,
                customRedirects: customRedirects,
                showPreviewiOS: showPreviewIos,
                showPreviewAndroid: showPreviewAndroid,
                copyToClipboardiOS: copyToClipboardIos,
                copyToClipboardAndroid: copyToClipboardAndroid,
                trackingCampaign: trackingCampaign,
                trackingSource: trackingSource,
                trackingMedium: trackingMedium
            ) { url in
                if let url = url {
                    result(url.absoluteString)
                } else {
                    result(FlutterError(code: "GENERATION_ERROR", message: "Failed to generate link", details: nil))
                }
            }
            
        case "setPushToken":
            guard let args = call.arguments as? [String: Any],
                  let token = args["token"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "token is required", details: nil))
                return
            }
            
            Grovs.pushToken = token
            result(nil)
            
        case "numberOfUnreadMessages":
            Grovs.numberOfUnreadMessages { count in
                result(count)
            }
            
        case "displayMessages":
            Grovs.displayMessagesViewController() {
                result(nil)
            }
            
        case "setUserIdentifier":
            guard let args = call.arguments as? [String: Any],
                  let identifier = args["identifier"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "identifier is required", details: nil))
                return
            }
            
            Grovs.userIdentifier = identifier
            result(nil)
            
        case "setUserAttributes":
            guard let args = call.arguments as? [String: Any],
                  let attributes = args["attributes"] as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "attributes are required", details: nil))
                return
            }
            
            Grovs.userAttributes = attributes
            result(nil)
            
        case "setDebugLevel":
            guard let args = call.arguments as? [String: Any],
                  let level = args["level"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "level is required", details: nil))
                return
            }
            
            let debugLevel: LogLevel
            switch level.lowercased() {
            case "info":
                debugLevel = .info
            case "error":
                debugLevel = .error
            default:
                debugLevel = .error
            }
            
            Grovs.setDebug(level: debugLevel)
            result(nil)
            
        case "logInAppPurchase":
            guard let args = call.arguments as? [String: Any],
                  let transactionIdString = args["transactionId"] as? String,
                  let transactionId = UInt64(transactionIdString) else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "transactionId is required", details: nil))
                return
            }

            Grovs.logInAppPurchase(transactionID: transactionId) { success in
                if success {
                    result(nil)
                } else {
                    result(FlutterError(code: "PAYMENT_ERROR", message: "Failed to log in-app purchase", details: nil))
                }
            }

        case "logCustomPurchase":
            guard let args = call.arguments as? [String: Any],
                  let typeString = args["type"] as? String,
                  let priceInCents = args["priceInCents"] as? Int,
                  let currency = args["currency"] as? String,
                  let productId = args["productId"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "type, priceInCents, currency, and productId are required", details: nil))
                return
            }

            let type: TransactionType
            switch typeString {
            case "buy":
                type = .buy
            case "cancel":
                type = .cancel
            case "refund":
                type = .refund
            default:
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid transaction type: \(typeString)", details: nil))
                return
            }

            var startDate: Date?
            if let dateString = args["startDate"] as? String {
                let formatter = ISO8601DateFormatter()
                startDate = formatter.date(from: dateString)
            }

            Grovs.logCustomPurchase(type: type, priceInCents: priceInCents, currency: currency, productID: productId, startDate: startDate) { success in
                if success {
                    result(nil)
                } else {
                    result(FlutterError(code: "PAYMENT_ERROR", message: "Failed to log custom purchase", details: nil))
                }
            }

        // Native analytics calls are fire-and-forget (void, non-throwing); nothing to surface to Flutter.
        case "track":
            guard let args = call.arguments as? [String: Any],
                  let name = args["name"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "name is required", details: nil))
                return
            }

            let properties = args["properties"] as? [String: Any]
            let tags = args["tags"] as? [String]
            Grovs.track(name, properties: properties, tags: tags)
            result(nil)

        case "setGlobalTags":
            let args = call.arguments as? [String: Any]
            let tags = args?["tags"] as? [String]
            Grovs.setGlobalTags(tags)
            result(nil)

        case "trackScreenView":
            guard let args = call.arguments as? [String: Any],
                  let screenName = args["screenName"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "screenName is required", details: nil))
                return
            }

            let properties = args["properties"] as? [String: Any]
            Grovs.trackScreenView(screenName, properties: properties)
            result(nil)

        case "setScreenAliases":
            guard let args = call.arguments as? [String: Any],
                  let aliases = args["aliases"] as? [String: String] else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "aliases is required", details: nil))
                return
            }

            Grovs.setScreenAliases(aliases)
            result(nil)

        case "setSDK":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "enabled is required", details: nil))
                return
            }

            setSDKEnabled(enabled)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

// MARK: - GrovsDelegate
extension GrovsPlugin: GrovsDelegate {
    public func grovsDidEncounterError(_ error: GrovsError, message: String) {
        errorStream.send(["code": error.description, "message": message])
    }

    public func grovsReceivedPayloadFromDeeplink(link: String?, payload: [String : Any]?, tracking: [String : Any]?) {
        // Drop lookups that complete after consent is withdrawn.
        guard GrovsPlugin.sdkEnabled else { return }
        guard let eventSink = eventSink else { return }
        
        var eventData: [String: Any] = [:]
        if let link = link {
            eventData["link"] = link
        }
        if let payload = payload {
            eventData["data"] = payload
        }
        if let tracking = tracking {
            eventData["tracking"] = tracking
        }
        
        eventSink(eventData)
    }
}

// MARK: - FlutterStreamHandler
extension GrovsPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/io/grovs/wrapper/GrovsPlugin.kt
package io.grovs.wrapper

import android.app.Activity
import android.app.Application
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import io.grovs.Grovs
import io.grovs.model.CustomLinkRedirect
import io.grovs.model.LogLevel
import io.grovs.model.exceptions.GrovsException
import io.grovs.service.CustomRedirects
import io.grovs.service.TrackingParams
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import io.grovs.model.events.PaymentEventType
import java.io.Serializable
import java.lang.ref.WeakReference
import java.util.Collections
import java.util.WeakHashMap

/** GrovsPlugin */
class GrovsPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private var channel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var errorChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var errorSink: EventChannel.EventSink? = null
    private var activityBinding: ActivityPluginBinding? = null
    private lateinit var coroutineScope: CoroutineScope
    private var newIntentListener: PluginRegistry.NewIntentListener? = null
    private val pendingDeeplinks = mutableListOf<Map<String, Any?>>()
    private val maxPendingDeeplinks = 20

    companion object {
        // Native configure and lifecycle registration must run once per process,
        // even when several Flutter engines attach.
        private var configured = false

        // Last consent value handed to the native SDK, shared by all engines.
        private var sdkEnabled = true
        private var consentGeneration = 0L
        private val attachedPlugins = Collections.newSetFromMap(WeakHashMap<GrovsPlugin, Boolean>())
        private var pendingIntent: Intent? = null

        private fun handleActivityStart(activity: Activity) {
            if (sdkEnabled) {
                pendingIntent?.let { activity.intent = it }
                pendingIntent = null
            }
            Grovs.onStart(activity)
        }
    }

    private val applicationLifecycleObserver: Application.ActivityLifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
        override fun onActivityCreated(p0: Activity, p1: Bundle?) { }
        override fun onActivityStarted(activity: Activity) {
            if (activity is FlutterActivity) {
                handleActivityStart(activity)
            }
        }
        override fun onActivityResumed(activity: Activity) { }
        override fun onActivityPaused(activity: Activity) { }
        override fun onActivityStopped(activity: Activity) { }
        override fun onActivitySaveInstanceState(activity: Activity, p1: Bundle) {}
        override fun onActivityDestroyed(activity: Activity) { }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
        attachedPlugins.add(this)
        
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "grovs")
        channel?.setMethodCallHandler(this)
        
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "grovs/deeplinks")
        eventChannel?.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
                if (events == null) return
                val pending = pendingDeeplinks.toList()
                pendingDeeplinks.clear()
                val generation = consentGeneration
                pending.forEach { deliverDeeplink(it, generation) }
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })

        errorChannel = EventChannel(flutterPluginBinding.binaryMessenger, "grovs/errors")
        errorChannel?.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                errorSink = events
            }

            override fun onCancel(arguments: Any?) {
                errorSink = null
            }
        })

        configureOnce(flutterPluginBinding.applicationContext as Application)
    }

    private fun configureOnce(application: Application) {
        if (configured) return
        val meta = application.packageManager
            .getApplicationInfo(application.packageName, PackageManager.GET_META_DATA)
            .metaData ?: Bundle()
        val apiKey = meta.getString("grovs_api_key") ?: ""
        val useTestEnvironment = meta.getBoolean("grovs_use_test_environment", false)
        val baseURL = meta.getString("grovs_base_url")
        val clipboardDomains = meta.getString("grovs_clipboard_domains")
            ?.split(',')
            ?.map { it.trim() }
            ?.filter { it.isNotEmpty() }
        sdkEnabled = meta.getBoolean("grovs_enabled", true)

        Grovs.configure(
            application,
            apiKey,
            useTestEnvironment,
            baseURL,
            false,
            clipboardDomains,
            sdkEnabled
        )
        application.registerActivityLifecycleCallbacks(applicationLifecycleObserver)
        configured = true
    }

    private fun setSDKEnabled(enabled: Boolean) {
        if (enabled == sdkEnabled) return
        Grovs.setSDK(enabled)
        sdkEnabled = enabled
        if (!enabled) {
            consentGeneration++
            attachedPlugins.forEach { it.pendingDeeplinks.clear() }
        }
        if (enabled) {
            // Re-run intent handling so a launch link and first-open attribution
            // that were skipped while disabled are processed now.
            activityBinding?.activity?.let { handleActivityStart(it) }
        }
    }

    private fun registerNativeDeeplinkListener(activity: Activity) {
        val plugin = WeakReference(this)
        val scope = coroutineScope
        Grovs.setOnDeeplinkReceivedListener(activity) { linkDetails ->
            if (!sdkEnabled) return@setOnDeeplinkReceivedListener
            val generation = consentGeneration
            scope.launch {
                plugin.get()?.deliverDeeplink(
                    mapOf(
                        "link" to linkDetails.link,
                        "data" to linkDetails.data,
                        "tracking" to linkDetails.tracking
                    ),
                    generation
                )
            }
        }
    }

    private fun deliverDeeplink(event: Map<String, Any?>, generation: Long) {
        if (!sdkEnabled || generation != consentGeneration) return
        val sink = eventSink
        if (sink != null) {
            sink.success(event)
            return
        }
        pendingDeeplinks.add(event)
        if (pendingDeeplinks.size > maxPendingDeeplinks) {
            pendingDeeplinks.removeAt(0)
        }
    }

    private fun attachActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        registerNativeDeeplinkListener(binding.activity)

        val listener = PluginRegistry.NewIntentListener { intent ->
            if (!sdkEnabled) pendingIntent = intent
            Grovs.onNewIntent(intent, binding.activity)
            false
        }
        newIntentListener = listener
        binding.addOnNewIntentListener(listener)
    }

    private fun detachActivity() {
        newIntentListener?.let { activityBinding?.removeOnNewIntentListener(it) }
        newIntentListener = null
        activityBinding = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            
            "generateLink" -> {
                val title = call.argument<String>("title")
                val subtitle = call.argument<String>("subtitle")
                val imageURL = call.argument<String>("imageURL")
                val data = call.argument<Map<String, Any>>("data")
                val tags = call.argument<List<String>>("tags")
                val customRedirectsMap = call.argument<Map<String, Any>>("customRedirects")
                val showPreviewIos = call.argument<Boolean>("showPreviewIos")
                val showPreviewAndroid = call.argument<Boolean>("showPreviewAndroid")
                val copyToClipboardIos = call.argument<Boolean>("copyToClipboardIos")
                val copyToClipboardAndroid = call.argument<Boolean>("copyToClipboardAndroid")
                val trackingMap = call.argument<Map<String, Any>>("tracking")
                
                if (title == null) {
                    result.error("INVALID_ARGUMENT", "title is required", null)
                    return
                }
                
                // Parse customRedirects
                val customRedirects = customRedirectsMap?.let { redirectsMap ->
                    val ios = (redirectsMap["ios"] as? Map<*, *>)?.let { iosMap ->
                        CustomLinkRedirect(
                            link = iosMap["url"] as? String ?: "",
                            openAppIfInstalled = iosMap["openAppIfInstalled"] as? Boolean ?: true
                        )
                    }
                    val android = (redirectsMap["android"] as? Map<*, *>)?.let { androidMap ->
                        CustomLinkRedirect(
                            link = androidMap["url"] as? String ?: "",
                            openAppIfInstalled = androidMap["openAppIfInstalled"] as? Boolean ?: true
                        )
                    }
                    val desktop = (redirectsMap["desktop"] as? Map<*, *>)?.let { desktopMap ->
                        CustomLinkRedirect(
                            link = desktopMap["url"] as? String ?: "",
                            openAppIfInstalled = desktopMap["openAppIfInstalled"] as? Boolean ?: true
                        )
                    }
                    CustomRedirects(ios = ios, android = android, desktop = desktop)
                }
                
                // Parse tracking params
                val tracking = trackingMap?.let { trackingParams ->
                    TrackingParams(
                        utmCampaign = trackingParams["utm_campaign"] as? String,
                        utmSource = trackingParams["utm_source"] as? String,
                        utmMedium = trackingParams["utm_medium"] as? String
                    )
                }
                
                // Convert data map to Serializable map
                val serializableData = data?.mapValues { entry ->
                    when (val value = entry.value) {
                        is Serializable -> value
                        is Number -> value
                        is String -> value
                        is Boolean -> value
                        else -> value.toString()
                    }
                }
                
                coroutineScope.launch {
                    try {
                        val link = withContext(Dispatchers.IO) {
                            Grovs.generateLink(
                                title = title,
                                subtitle = subtitle,
                                imageURL = imageURL,
                                data = serializableData,
                                tags = tags,
                                customRedirects = customRedirects,
                                showPreviewIos = showPreviewIos,
                                showPreviewAndroid = showPreviewAndroid,
                                copyToClipboardIos = copyToClipboardIos,
                                copyToClipboardAndroid = copyToClipboardAndroid,
                                tracking = tracking
                            )
                        }
                        result.success(link)
                    } catch (e: CancellationException) {
                        throw e
                    } catch (e: GrovsException) {
                        result.error("GENERATION_ERROR", e.message, null)
                    } catch (e: Exception) {
                        result.error("GENERATION_ERROR", e.message, null)
                    }
                }
            }
            
            "setPushToken" -> {
                val token = call.argument<String>("token")
                
                if (token == null) {
                    result.error("INVALID_ARGUMENT", "token is required", null)
                    return
                }
                
                try {
                    Grovs.pushToken = token
                    result.success(null)
                } catch (e: Exception) {
                    result.error("TOKEN_ERROR", e.message, null)
                }
            }
            
            "setUserIdentifier" -> {
                val identifier = call.argument<String>("identifier")
                
                if (identifier == null) {
                    result.error("INVALID_ARGUMENT", "identifier is required", null)
                    return
                }
                
                try {
                    Grovs.identifier = identifier
                    result.success(null)
                } catch (e: Exception) {
                    result.error("USER_ERROR", e.message, null)
                }
            }
            
            "setUserAttributes" -> {
                val attributes = call.argument<Map<String, Any>>("attributes")
                
                if (attributes == null) {
                    result.error("INVALID_ARGUMENT", "attributes are required", null)
                    return
                }
                
                try {
                    Grovs.attributes = attributes
                    result.success(null)
                } catch (e: Exception) {
                    result.error("USER_ERROR", e.message, null)
                }
            }
            
            "setDebugLevel" -> {
                val level = call.argument<String>("level")
                
                if (level == null) {
                    result.error("INVALID_ARGUMENT", "level is required", null)
                    return
                }
                
                try {
                    // Convert string to Grovs debug level
                    // Note: You may need to adjust this based on the actual Grovs SDK API
                    when (level.lowercase()) {
                        "info" -> Grovs.setDebug(LogLevel.INFO)
                        "error" -> Grovs.setDebug(LogLevel.ERROR)
                        else -> Grovs.setDebug(LogLevel.ERROR)
                    }
                    result.success(null)
                } catch (e: Exception) {
                    result.error("DEBUG_ERROR", e.message, null)
                }
            }
            
            "logInAppPurchase" -> {
                val transactionId = call.argument<String>("transactionId")

                if (transactionId == null) {
                    result.error("INVALID_ARGUMENT", "transactionId is required", null)
                    return
                }

                try {
                    Grovs.logInAppPurchase(transactionId)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("PAYMENT_ERROR", e.message, null)
                }
            }

            "logCustomPurchase" -> {
                val typeString = call.argument<String>("type")
                val priceInCents = call.argument<Int>("priceInCents")
                val currency = call.argument<String>("currency")
                val productId = call.argument<String>("productId")
                val startDateString = call.argument<String>("startDate")

                if (typeString == null || priceInCents == null || currency == null || productId == null) {
                    result.error("INVALID_ARGUMENT", "type, priceInCents, currency, and productId are required", null)
                    return
                }

                val type = when (typeString) {
                    "buy" -> PaymentEventType.BUY
                    "cancel" -> PaymentEventType.CANCEL
                    "refund" -> PaymentEventType.REFUND
                    else -> {
                        result.error("INVALID_ARGUMENT", "Invalid transaction type: $typeString", null)
                        return
                    }
                }

                try {
                    Grovs.logCustomPurchase(type, priceInCents, currency, productId)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("PAYMENT_ERROR", e.message, null)
                }
            }

            "track" -> {
                val name = call.argument<String>("name")
                val properties = call.argument<Map<String, Any>>("properties")
                val tags = call.argument<List<String>>("tags")

                if (name == null) {
                    result.error("INVALID_ARGUMENT", "name is required", null)
                    return
                }

                try {
                    Grovs.track(name, properties, tags)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("EVENT_ERROR", e.message, null)
                }
            }

            "setGlobalTags" -> {
                val tags = call.argument<List<String>>("tags")
                try {
                    Grovs.setGlobalTags(tags)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("EVENT_ERROR", e.message, null)
                }
            }

            "trackScreenView" -> {
                val screenName = call.argument<String>("screenName")
                val properties = call.argument<Map<String, Any>>("properties")

                if (screenName == null) {
                    result.error("INVALID_ARGUMENT", "screenName is required", null)
                    return
                }

                try {
                    Grovs.trackScreenView(screenName, properties)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("EVENT_ERROR", e.message, null)
                }
            }

            "setScreenAliases" -> {
                val aliases = call.argument<Map<String, String>>("aliases")

                if (aliases == null) {
                    result.error("INVALID_ARGUMENT", "aliases is required", null)
                    return
                }

                try {
                    Grovs.setScreenAliases(aliases)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("EVENT_ERROR", e.message, null)
                }
            }

            "setSDK" -> {
                val enabled = call.argument<Boolean>("enabled")

                if (enabled == null) {
                    result.error("INVALID_ARGUMENT", "enabled is required", null)
                    return
                }

                try {
                    setSDKEnabled(enabled)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("CONSENT_ERROR", e.message, null)
                }
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        eventChannel?.setStreamHandler(null)
        errorChannel?.setStreamHandler(null)
        // Removing stream handlers does not call onCancel.
        eventSink = null
        errorSink = null
        pendingDeeplinks.clear()
        detachActivity()
        coroutineScope.cancel()
        attachedPlugins.remove(this)
        channel = null
        eventChannel = null
        errorChannel = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        attachActivity(binding)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        detachActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        attachActivity(binding)
    }

    override fun onDetachedFromActivity() {
        detachActivity()
    }
}
